contracts/yellow/VaultImpl.sol

```diff
@@ -1,11 +1,266 @@
+     * @param brokerKeyDerivedAddress_ Address derived from Broker public key.
+     * @param otpKeyDerivedAddress_ Address derived from OTP public key.
+     */
+    constructor(address brokerKeyDerivedAddress_, address otpKeyDerivedAddress_) {
+        _brokerKeyDerivedAddress = brokerKeyDerivedAddress_;
+        _otpKeyDerivedAddress = otpKeyDerivedAddress_;
+    }
+
+    /**
+     * @notice Get last ledger id (deposits and withdrawals id).
+     * @return uint256 Ledger id.
+     */
+    function getLastId() external view override returns (uint256) {
+        return _ledgerId.current();
+    }
+
+    /**
+     * @notice Set the address derived from the broker's new public key. Emits `BrokerKeyDerivedAddressSet` event.
+     * @dev Supplied payload must be signed by broker's current public key.
+     * @param encodedAddress Encoded address to set.
+     * @param signature Encoded address signed by broker's current public key.
+     */
+    function setBrokerKeyDerivedAddress(bytes calldata encodedAddress, bytes calldata signature) external {
+      bytes32 digest = ECDSA.toEthSignedMessageHash(keccak256(encodedAddress));
+      address recoveredBrokerDerivedKey = ECDSA.recover(digest, signature);
+      require(recoveredBrokerDerivedKey == _brokerKeyDerivedAddress, 'Vault: signer is not broker');
+      _brokerKeyDerivedAddress = recoveredBrokerDerivedKey;
+
+      emit BrokerKeyDerivedAddressSet(recoveredBrokerDerivedKey);
+    }
+
+    /**
+     * @notice Set the address derived from the OTP's new public key. Emits `OTPKeyDerivedAddressSet` event.
+     * @dev Supplied payload must be signed by OTP's current public key.
+     * @param encodedAddress Encoded address to set.
+     * @param signature Encoded address signed by OTP's current public key.
+     */
+    function setOTPKeyDerivedAddress(bytes calldata encodedAddress, bytes calldata signature) external {
+      bytes32 digest = ECDSA.toEthSignedMessageHash(keccak256(encodedAddress));
+      address recoveredOTPDerivedKey = ECDSA.recover(digest, signature);
+      require(recoveredOTPDerivedKey == _otpKeyDerivedAddress, 'Vault: signer is not otp');
+      _otpKeyDerivedAddress = recoveredOTPDerivedKey;
+
+      emit OTPKeyDerivedAddressSet(recoveredOTPDerivedKey);
+    }
+
+    /**
+     * @notice Deposit assets with given payload from the caller. Emits `Deposited` event.
+     * @param payload Encoded payload, which consists of rid (unique identifier id), expire timestamp, deposit address and an array of allocations.
+     * @param brokerSignature Payload signed by the Broker.
+     * @param otpSignature Payload signed by the OTP service.
+     * @return bool Return 'true' if deposited successfully.
+     */
+    function deposit(
+        bytes calldata payload,
+        bytes calldata brokerSignature,
+        bytes calldata otpSignature
+    ) external payable returns (bool) {
+        _requireValidInput(DEPOSIT_TYPE, payload, brokerSignature, otpSignature);
```

Comment:
I am not sure we need `DEPOSIT_TYPE` and `WITHDRAW_TYPE` as they are manually set in the functions.
If we want to add another point of security, we may supply those types in payload itself. Nevertheless, this would doubtfully be a security increase.

What do you think, @lckmidnight ?

Reply:
Decided to add the type into `payload`.